### ISSUE TITLE
watch keeps being triggered while making a webpackPlugin

### ISSUE THREAD

**Issue by flyon** [TypeStrong/ts-loader#137]
Hey, this probably has to do with the fact that I'm not good enough yet with making webpackPlugins, but I'm starting to expect the issue I'm having is connected to what triggers a rebuild in ts-loader, so I was hoping you could help.

I'm using the new --declaration option together with --watch and my own plugin. The plugin takes all generated .d.ts files, moves them to the `/tmp` folder and then generates a single declaration file and saves that in the `/build` folder. All the source, including the entry point is located in `/src`. 

The thing is that the first compile seems to go OK, but once the watch get's triggered on a new change, things go into a loop. If I do not move the generated .d.ts files it doesnt go looping. 

So in other words, it seems the watcher is triggered if I remove/move the generated declaration files in the /src folder. But why is it doing that? Nothing in the sourcecode is actually using / referencing those temporary declaration files if you follow the source from the entry point. 

I tried changing the loader in order not to match .d.ts, but that didnt help.

            { test: /(?!.*\.d.ts)\.ts(x?)$/, loader: 'ts-loader' }

any ideas?

I could ofcourse just keep the .d.ts files around, but it just makes all the source folders a bit more bloated with partial .d.ts files that I'm not using, so I'd like to be able to remove (without triggering loops) if possible.

**Comment by jbrantly** [comment]
That's interesting. Without actually investigating, I have two ideas.

1) That webpack for some reason watches files that are emitted from the loader with `emitFile`. That just seems weird though, my guess is that it's not doing this.

2) ts-loader [has code](https://github.com/TypeStrong/ts-loader/blob/master/index.ts#L556-L562) that adds a dependency on all .d.ts files for each .ts file that goes through the loader. If your `tsconfig.json` does not include an explicit `files` array, it's possible that all of those .d.ts files are being added to TypeScript (on subsequent builds) and thus webpack is watching them.

**Comment by flyon** [comment]
thank you for thinking along

yes 2) sound likely. I do have `files` specified in `tsconfig.json`, it only contains one file `src/init.ts`, which is the same as the `entry` specified in webpack.config.js.

having files specified should in theory mean the .d.ts should not be added to the files being watched?

**Comment by jbrantly** [comment]
In theory yes. Might be worth adding a couple of logging statements to see what tsloader is adding through `addDependency`

**Comment by flyon** [comment]
I added logging to see which declarationFiles are going through `allDefinitionFiles.forEach(this.addDependency.bind(this));` and it looks like the new declaration files generated because of `--declaration` are _not_ added as dependency

**Comment by flyon** [comment]
compiler.plugin('done', (stats) => {
        console.log(stats.compilation.fileTimestamps);

after the first compilation this is an empty Object, after the second compilation it contains all the dependencies that were added by ts-loader, but __also__ the generated .d.ts files.

So it looks a lot like your suggested in 1) 

I'm thinking I'll probably have to learn a bit more about the workings of webpack and see if I can make the plugin in such a way that the generated .d.ts files are removed from the files to emit before they are actually emited.

Closing because it looks like it's not a ts-loader issue, tnx for the help!

**Comment by jbrantly** [comment]
Good luck! If you get it working let me know so I can promote it here.

**Comment by flyon** [comment]
I got it working! .. by switching to listening for 'emit' instead of 'done, I was able to remove the declaration files from the list of files that are to be emited, merge them, and add back the merged 1 file to the list of files to emit. (much like the example [here](https://github.com/webpack/docs/wiki/how-to-write-a-plugin))

I've published it on NPM as [dacore-webpack-declarations](https://www.npmjs.com/package/dacore-webpack-declarations), but my usecase is quite specific, since I do something quite 'stange', I change the declarations so that it all looks like an internal module. 

I may see later if I can make it more general applicable. For now, it is at most a good starting point if someone wants to make his own plugin.

**Comment by flyon** [comment]
I've made it more general, I tried makign it applicable to the general case of bundling external module declaration files, but found out that requires some sort of module naming scheme based on the file names. Which will already be included in Typescript 1.8.

So for now it is configurable and usable for other who also wish to write external modules but consume them as internal modules, which probably isnt a lot of people.

Might still serve as inspiration for another plugin though..
